### PR TITLE
Add net.cozic.joplin_desktop

### PR DESCRIPTION
**Pull request by catsout**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/laurent22/joplin/issues/738
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by catsout** [comment]
- This build now link my fork of flatpak-builder-tools, as it need modification, will create pr later.
- ID is `org.joplinapp.Joplin` or `net.cozic.joplin-desktop`?
- How to support open external markdown editor?
- Is there a good way to add support to system-tray

**Comment by catsout** [comment]
bot, build org.joplinapp.Joplin

**Comment by catsout** [comment]
bot, build org.joplinapp.Joplin

**Comment by tinywrkb** [comment]
> Is there a good way to add support to system-tray?

`- --own-name=org.kde.*` should solve this but when I tried (my own packaging, not this) Joplin didn't create a StatusNotifierItem so while an icon appears and the menu in the status bar is working, the icon itself is blank. Maybe this is a bug?

> How to support open external markdown editor?

`flatpak-spawn --host` might work if the target app gets the correct path to file in `xdg-config/joplin-desktop/edit-<UUID>.md`, and if it's Flatpak then it also needs filesystem access permission. I believe you need  `--talk-name=org.freedesktop.Flatpak` to open apps on the host.

**Comment by catsout** [comment]
> `- --own-name=org.kde.*` should solve this but when I tried (my own packaging, not this) Joplin didn't create a StatusNotifierItem so while an icon appears and the menu in the status bar is working, the icon itself is blank. Maybe this is a bug?
No, check by dbus-monitor, the icon dir seems to be put at /tmp/.org.chromium.Chromium* in sandbox.
Set TMPDIR fixed.

**Comment by tinywrkb** [comment]
> Set TMPDIR=$XDG_CACHE_HOME fixed

Use instead `export TMPDIR=$XDG_RUNTIME_DIR/app/$FLATPAK_ID` and put this in your `joplin-desktop.sh`.

**Comment by tinywrkb** [comment]
Here's an example of how to get an external editor working.

Path: `/bin/flatpak-spawn`  
Arguments: `--host /bin/flatpak run --filesystem=xdg-config/joplin-desktop org.gnome.gedit`

**Comment by catsout** [comment]
> flatpak-spawn --host might work if the target app gets the correct path to file in xdg-config/joplin-desktop/edit-<UUID>.md, and if it's Flatpak then it also needs filesystem access permission. I believe you need --talk-name=org.freedesktop.Flatpak to open apps on the host.

Thank for this, it works, seems like need a readme for it.
I also try xdg-open, also ok, but need choose every time.

File: org.joplinapp.Joplin.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  - org.freedesktop.Sdk.Extension.node12
+separate-locales: false
+command: joplin-desktop
+finish-args:
+  - --socket=pulseaudio
+  - --socket=fallback-x11
+  - --device=dri
+  - --share=ipc
+  - --share=network
+  - --filesystem=xdg-config/joplin-desktop:create
```
[the remark this answers is not in the document]
Comment (catsout) [reply org.joplinapp.Joplin.yml:17]: Actually, I set xdg-config to let it use existing notes, as there is no problem to mess existing notes.  
Also, if use persist, there will be a problem that not see existing notes for users who switch to flatpak version.

**Comment by catsout** [comment]
@laurent22
Any suggestions about the id, build script, metainfo(will display on flathub app page) and so on ?

**Comment by barthalion** [comment]
LGTM (other than metadata/desktop but you seem to be aware of that). @tinywrkb is it something you would like to co-maintain if @catsout does not mind?

File: org.joplinapp.Joplin.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jurf) [reply org.joplinapp.Joplin.yml:17]: TBH I’d prefer persist, one of the great advantages of Flatpak is that it doesn’t pollute the homedir. I’m not sure I’m aware of any app that does this but correct me if I’m wrong.

**Comment by jurf** [comment]
Thanks for the build, I just started using this yesterday, incredible timing.

**Comment by catsout** [comment]
> LGTM (other than metadata/desktop but you seem to be aware of that). @tinywrkb is it something you would like to co-maintain if @catsout does not mind?

It's okay for me.

**Comment by catsout** [comment]
Joplin support import and output md,evernote,config files.
So, seems that we should add `filesystem=xdg-documents`?

**Comment by barthalion** [comment]
I think you can attach any file so I'd consider just `--filesystem=home`. It is what it is, you can't really get it right without using the portal.

**Comment by laurent22** [comment]
> @laurent22
> Any suggestions about the id, build script, metainfo(will display on flathub app page) and so on ?

The desktop app ID is "net.cozic.joplin-desktop" so for consistency I'd use that (the mobile and cli apps are similarly named). For the build script, I'm not familiar with flatpak so can't help much but if there's something in particular not working, please let me know.

For metainfo, there's no much in the app package.json, so I guess you could take it from the main website.

**Comment by catsout** [comment]
bot, build net.cozic.joplin_desktop

**Comment by catsout** [comment]
> The desktop app ID is "net.cozic.joplin-desktop" so for consistency I'd use that 

As [flatpak doc](https://docs.flatpak.org/en/latest/conventions.html#application-ids), It should use "net.cozic.joplin_desktop" in flatpak.

**Comment by tinywrkb** [comment]
> LGTM (other than metadata/desktop but you seem to be aware of that). @tinywrkb is it something you would like to co-maintain if @catsout does not mind?

I don't mind helping to maintain the package but I have zero knowledge about Node.js so I'm not sure how much I will be able to contribute.

**Comment by jurf** [comment]
> I think you can attach any file so I'd consider just `--filesystem=home`. It is what it is, you can't really get it right without using the portal.

Maybe `--filesystem=host` to include external devices? It’s been a while since I’ve looked into this, however.

**Comment by rhamzeh** [comment]
> > I think you can attach any file so I'd consider just `--filesystem=home`. It is what it is, you can't really get it right without using the portal.
> 
> Maybe `--filesystem=host` to include external devices? It’s been a while since I’ve looked into this, however.

I know Flatpak is being used here for portability and packaging, but let's try to keep the permissions required as minimal as possible by default, and *document* what users need to do if they want anything more than the defaults.

I would argue that both filesystem=home and filesystem=host are a bit exessive?

Maybe keep `xdg-documents`, `xdg-downloads` as well as `/media` (Documents + Downloads + External Storage) to cover a big chunk of the use cases, without going all the way of granting access to everything in the system?

It can be document how to add and remove permissions to make it easier for users.
e.g. The Readme/Wiki can say:
If you want to export or import data from a directory, you need to grant Joplin access to it first by running the following command, then restarting Joplin:
`flatpak override net.cozic.joplin-desktop --filesystem=<path_here>`

**Comment by barthalion** [comment]
It's always hard to strike the balance with Electron apps. If not `home`, the bare minimum would be all `xdg-*` permissions AND `/run/media` + `/media`. However, as it already has access to `flatpak-spawn` for supporting external editors, the sandbox mostly acts as efficient packaging medium and not a real sandbox. The difference between `home` and `host` is in practice minimal here.

**Pull request closed without merging**